### ISSUE TITLE
Implement focus stealing prevention

### ISSUE THREAD

**Issue by Saviq** [canonical/mir#2586]
I'm running Thunderbird under Miriway (same issue with notifications from websites in Firefox). The notification surfaces (part of the problem is lack of a notification manager) show in random locations on the screen, incorrectly cropped and steal focus.

There may be nothing that Mir can do visually, if those are simply top-level surfaces - but focus stealing prevention is something we need to do. May need a separate issue?

Using this [Miriway](https://github.com/Miriway/Miriway) config:

```ini
x11-window-title=Miriway
idle-timeout=180
ctrl-alt=t:miriway-terminal # Default "terminal emulator finder"

enable-x11=true

shell-component=swaybg -i /usr/share/backgrounds/warty-final-ubuntu.png
shell-component=mate-panel
meta=a:mate-panel --run-dialog
ctrl-alt=t:mate-terminal
```

**Comment by wmww** [comment]
We *might* be able to do something to prevent focus stealing, but fwiw the problem exists in sway as well. Easiest solution is probably to run mate-notification-daemon, which I ported to layer shell a while back.

**Comment by Saviq** [comment]
> Easiest solution is probably to run mate-notification-daemon, which I ported to layer shell a while back.

Looks like it wants to be DBus-activated, as it silently exits a few seconds after starting...

**Comment by Saviq** [comment]
> Looks like it wants to be DBus-activated, as it silently exits a few seconds after starting...

Hmm, it even _is_ DBus-activated I think. Still shows up in the middle of the screen, though.
![frame_2022-08-26T10:51+02:00](https://user-images.githubusercontent.com/231077/186867741-c87dc6bf-0072-4da2-a788-967a841fde76.png)

**Comment by AlanGriffiths** [comment]
> > Looks like it wants to be DBus-activated, as it silently exits a few seconds after starting...
> 
> Hmm, it even _is_ DBus-activated I think. Still shows up in the middle of the screen, though. ![frame_2022-08-26T10:51+02:00](https://user-images.githubusercontent.com/231077/186867741-c87dc6bf-0072-4da2-a788-967a841fde76.png)

It won't have access to `zwlr_layer_shell` unless started from `miriway-shell`, so try enabling `add-wayland-extensions=all` and see if that works

**Comment by Saviq** [comment]
> It won't have access to `zwlr_layer_shell` unless started from `miriway-shell`, so try enabling `add-wayland-extensions=all` and see if that works

Bingo!

I've modified the bug to talk about focus stealing prevention only.

**Comment by AlanGriffiths** [comment]
> Bingo!

Well, assuming we don't want to allow zwlr_layer_shell _et alias_ to all applications Miriway needs a way to "whitelist" such services. (Might nor might not be a Mir issue too)

**Comment by tarek-y-ismail** [comment]
Quick reformulation of the issue: As we have xdg-activation implemented now, there's no need to automatically focus and bring new windows to the front. xdg-activation can then raise and focus windows when it's appropriate.

I've tracked what's needed down to two places, **these may not be the proper places to modify, but are a start**:

1. `MinimalWindowManager::handle_window_ready`. This part is responsible for focusing new windows.
https://github.com/canonical/mir/blob/3c9a7c64dbb99d273d82d5fe1b7f9daa092d2df8/src/miral/minimal_window_manager.cpp#L127-L133

2. `SurfaceStack::insert_surface_at_top_of_depth_layer`. This is responsible for the placement of a window within a depth layer.
https://github.com/canonical/mir/blob/3c9a7c64dbb99d273d82d5fe1b7f9daa092d2df8/src/server/scene/surface_stack.cpp#L513-L519

Modifying these two pieces of code to not automatically focus and push new windows to the front should allow xdg-activation to do so when it's appropriate.

What's missing: Creating and passing some sort of strategy down to control the behavior depending on whether xdg-activation is available or not. One approach could be using `mir::scene::Surface` to carry this information since it can be accessed through (1), and can be exposed via `WindowInfo -> WindowInfo::Self -> Window -> Window::operator std::shared_ptr<mir::scene::Surface>()` in (2) (though I think we don't have direct access to the surface in this case for a reason)

**Comment by tarek-y-ismail** [comment]
One thing I'm realizing re-reading [xdg-activation-v1](https://wayland.app/protocols/xdg-activation-v1) is that it only applies to applications launched from other applications.  So at the moment with the changes mentioned applying to _all_ applications, applications launched through shortcuts will not start focused (can be worked around)

**Comment by Saviq** [comment]
Opening links _is_ launching (or activating, if already running) applications from other applications. That just requires things to work together and pass the token further. 

This was actually _the_ use case Matt implemented activation for.

**Comment by tarek-y-ismail** [comment]
Oh I think I should've elaborated more. I didn't mean URLs when I mentioned shortcuts. I meant Mir's shortcuts (CTRL+ALT+T for example)

**Comment by AlanGriffiths** [comment]
> Oh I think I should've elaborated more. I didn't mean URLs when I mentioned shortcuts. I meant Mir's shortcuts (CTRL+ALT+T for example)

Yes, we need to update our launch code to supply a token we respect

**Comment by tarek-y-ismail** [comment]
Some things that broke with the changes above:
1. Since window offsets are based off the currently open window, only the second window is offset, and all windows launched after the second window are positioned at the exact same position.
2. Xterm (probably all server-decorated apps) are borked
![20241129_14h43m57s_grim](https://github.com/user-attachments/assets/0144c7c7-04d3-4ddb-8792-b7c145f0c2a9)

**Comment by tarek-y-ismail** [comment]
Spent a fair bit of time trying to figure out a way to propagate the information that `xdg_activation_v1` (or really, whatever extension needs to be checked) down to `BasicWindow` so it can be used in `handle_window_ready`.  This was done through exposing `Server::is_wayland_extension_enabled`, and adding `miral::ExtensionLookup` so it can be used in miral.

At the moment, I'm a bit lost how to do this with `ApplicationSession::create_surface` (check previous comment). The most obvious way after a quick glance of the stack trace and code would be to pass the info through `SessionManager`'s constructor, then through `ApplicationSession`'s constructor, then using it in `create_surface`.

**Comment by AlanGriffiths** [comment]
> Spent a fair bit of time trying to figure out a way to propagate the information that `xdg_activation_v1` (or really, whatever extension needs to be checked) down to `BasicWindow` so it can be used in `handle_window_ready`. This was done through exposing `Server::is_wayland_extension_enabled`, and adding `miral::ExtensionLookup` so it can be used in miral.

I think this approach is problematic: We have APIs that allow shells fine grained control over enabling extensions (c.f. `miral::WaylandExtensions::conditionally_enable()`, so there is no way to implement `Server::is_wayland_extension_enabled` _et alia_ to produce reliable results. Also, as you have noted, it requires significant work to propagate the results to where they would be useful.

I would reverse the dependency: if "focus stealing prevention" is enabled in miral, then force `xdg_activation_v1`to be enabled.

**Comment by AlanGriffiths** [comment]
> I would reverse the dependency: if "focus stealing prevention" is enabled in miral, then force `xdg_activation_v1`to be enabled.

Or even, just allow the options to be independent: even if not the desired end state, that makes experimentation easier

**Comment by Saviq** [comment]
> allow the options to be independent

That may well be the best — `--require-activation` or so, and no connection to the activation protocol?

**Comment by tarek-y-ismail** [comment]
Spent some time passing things (`ExtensionLookup`) to `ApplicationSession` so surfaces can be inserted at the proper place in the stack. I also spent some time trying out different applications just to have an idea which use xdg-activation-v1. 

After that, Alan, Sawicz, and I spent some time discussing whether I was going the right way and using Mir's existing customization options properly or not (hint: I was not). So, seeing how alt+tab did something similar to what we did, we agreed that this was a good place to start and figure out if `FloatingWindowManagerPolicy` needed any changes to expose that so it can be customized in it. The next step would've been to then add a new constructor for `FloatingWindowManagerPolicy` to toggle between the old and new logic.

I spent some time reading through c2553c5adf86b47e5199835be3d778d67bf0b4c3 to get a basic idea how alt + tab was implemented, figured that `WindowManagerTools::swap_tree_order` was the missing piece, and figured out what to do to disable auto raising (just swap the new window on top with the already existing one) + focusing (at least for all applications after the first, makes it easier to test :)). Was trolled a bit by kgx actually raising windows because I forgot to disable xdg-activation :P

As of right now, this is done with a few lines in `FloatingWindowManager`, but I believe that I discovered a bug? where the next window in the stack is not focused whenever we close the one above it. I'm not sure of the root cause at the moment.
